File: 0050-powx-n/0050-powx-n.js
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
    
    let pow = (x, n) => {
        if( n == 0) {
            return 1;
        }
        let temp = pow(x, parseInt(n/2));
        if(n%2 == 0) {
            return temp * temp;
        } else {
            return x * temp * temp;
        }
    }
    
    let value = pow(x, Math.abs(n));
    if(n >= 0) return value;
    return 1/value;
};

File: 0050-powx-n/0050-powx-n.java
class Solution {
    public HashMap<Integer, Double> map; // = new HashMap<int, double>();
    public double pow(double x, int n) {
        if(n == 0) {
            return 1;
        }
        if(map.containsKey(n)) {
            return map.get(n);
        }
        if(n%2 == 0) {
            map.put(n , this.pow(x, n / 2) * this.pow(x, n / 2));
        } else {
            map.put(n , x * this.pow(x, n / 2) * this.pow(x, n / 2));
        }
        return map.get(n);
    }
    public double myPow(double x, int n) {
       this.map = new HashMap<Integer, Double>();
       return n >= 0 ? pow(x,n) : (double) 1 / pow(x,n);
    }
}